tools/epoch-sync/src/cli.rs

```diff
@@ -0,0 +1,293 @@
+            tracing::info!(?snapshot_path, "Found a DB snapshot");
+        } else {
+            tracing::info!(destination = ?snapshot_path, "Creating snapshot of original DB");
+            // checkpointing only hot storage, because cold storage will not be changed
+            checkpoint_hot_storage_and_cleanup_columns(
+                &storage.get_hot_store(),
+                &snapshot_path,
+                None,
+            )?;
+        }
+
+        near_config.config.store.path = Some(store_path_addition.join("data"));
+
+        Ok(near_config)
+    }
+}
+
+#[derive(clap::Parser)]
+struct ValidateEpochSyncInfoCmd {
+    /// If `archive` flag is specified, `BlockInfo` column is assumed to be full and is used for optimisation purposes.
+    /// Using `BlockInfo` (`--archive` flag) takes around 10 minutes.
+    /// Using `BlockHeader` takes around 1.5 hours.
+    #[clap(short, long)]
+    archive: bool,
+}
+
+impl ValidateEpochSyncInfoCmd {
+    pub fn run(
+        &self,
+        home_dir: &Path,
+        storage: &NodeStorage,
+        config: &nearcore::config::NearConfig,
+    ) -> anyhow::Result<()> {
+        let store = storage.get_hot_store();
+
+        let hash_to_prev_hash = if self.archive {
+            get_hash_to_prev_hash_from_block_info(storage)?
+        } else {
+            get_hash_to_prev_hash_from_block_header(storage)?
+        };
+        let epoch_ids = get_all_epoch_ids(storage)?;
+
+        let mut chain_store =
+            ChainStore::new(store.clone(), config.genesis.config.genesis_height, false);
+        let header_head_hash = chain_store.header_head()?.last_block_hash;
+        let hash_to_next_hash = get_hash_to_next_hash(&hash_to_prev_hash, &header_head_hash)?;
+
+        let epoch_manager =
+            EpochManager::new_arc_handle(storage.get_hot_store(), &config.genesis.config);
+        let shard_tracker = ShardTracker::new(
+            TrackedConfig::from_config(&config.client_config),
+            epoch_manager.clone(),
+        );
+        let runtime = NightshadeRuntime::from_config(
+            home_dir,
+            storage.get_hot_store(),
+            &config,
+            epoch_manager.clone(),
+        );
+        let chain_update = ChainUpdate::new(
+            &mut chain_store,
+            epoch_manager,
+            shard_tracker,
+            runtime,
+            DoomslugThresholdMode::TwoThirds,
+            config.genesis.config.transaction_validity_period,
+        );
+
+        let genesis_hash = store
+            .get_ser::<CryptoHash>(
+                DBCol::BlockHeight,
+                &config.genesis.config.genesis_height.to_le_bytes(),
+            )?
+            .expect("Expect genesis height to be present in BlockHeight column");
+
+        let mut cur_hash = header_head_hash;
+
+        // Edge case if we exactly at the epoch boundary.
+        if epoch_ids.contains(&cur_hash) {
+            cur_hash = hash_to_prev_hash[&cur_hash];
```

Comment:
Why is it ok to skip the header head block?
Don't we need to check that the last block of an epoch exists and has block info?

Reply:
`EpochSyncInfo` is only constructed when the next epoch has a finalised block, because we need a header with `epoch_sync_data_hash` present.

Reply:
Thanks, please add that as a comment to the code.